neps/nep-0461.md

```diff
@@ -0,0 +1,72 @@
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/461
+Status: Review
+Type: Protocol
+Category: Runtime
+Version: 1.0.0
+Created: 14-Feb-2023
+Updated: 14-Feb-2023
+---
+
+## Summary
+A proposal to include a standardized prefix in all standardized signable messages, to distinguish actionable messages (i.e. those that are validated on chain), from non-actionable messages (i.e. those that are used off-chain), from [classic transactions](https://nomicon.io/RuntimeSpec/Transactions).
+
+Particularly, we propose to prepend the following prefixes:
+
+| Prefix | Message |
+| - | - |
+| $< 2^{30}$ | [Classic Transactions](https://nomicon.io/RuntimeSpec/Transactions) |  
+| $2^{30}$ + NEP-number  | On-chain Actionable Message, e.g. [`SignedDelegateAction`](https://github.com/near/NEPs/blob/master/neps/nep-0366.md) |
+| $2^{31}$ + NEP-number  | Off-chain Message, e.g. [`signMessage`](https://github.com/near/NEPs/pull/413) |
+
+## Motivation
+As the ecosystem grows, more standards will appear, enabling users to sign messages that are not transactions (e.g. [NEP413](https://github.com/near/NEPs/pull/413/), and [NEP366](https://github.com/near/NEPs/blob/master/neps/nep-0366.md)). If such messages collide with transactions, a malicious app could ask users to sign them, only to later relay them as valid transactions to the network.
+
+## Rationale and alternatives
+
+In NEAR, users sign transactions that are later relayed to the network. At the moment of being relayed, the transactions are simple streams of bytes that a node then parses into a [valid transaction](https://nomicon.io/RuntimeSpec/Transactions).
+
+As the protocol expands, more messages will be added for users to sign. We want users to be able to use the protocol safely, making sure that all those signed message cannot be interpreted as a valid transaction.
+
+Including a `u32` fixed prefix will enable to easily distinguish transactions from messages, while also ensuring that they never overlap. This is because the first parameter of a [valid transactions](https://nomicon.io/RuntimeSpec/Transactions) is an `AccountId`, which is encoded as `string length <u32>` + `string chars <u32[]>`. Adding a prefix in the range of $2^30 - 2^32$ will automatically make the non-transactional message to deserialize into an invalid transaction, since no `AccountId` has more than 64 chars.
+
+## Specification
+
+### On-Chain Messages
+All actionable messages, i.e. those that ARE validated on-chain, should have a structure which first parameter is a `prefix: u32` with fixed value $2^{30} + NEP-number$.
+
+For example, if we implement a new actionable message, destinated to be used on-chain, and the proposal has the number `NEP400`, then the structure must start with the prefix `1073742224`:
+
+```ts
+Structure{
+    prefix: u32 = 1073742224 // prefix fixed to 2**30 + 400 == 1073742224
+    // Rest of the attributes go here
+}
+```
```

Comment:
Two technical details:
1. This reads as if the prefix must be included in the message itself. In the implementation, it feels more natural to not include it in the actual message, only in the input for the signature.
2. Do you want to sign the hash of the message or the message itself?

More specifically, NEP-366 (meta transaction) defines signatures like this:
```rust
// NEP-366
let delegate_action = DelegateAction {...};
let signature = sign(hash(delegate_action));
let signed_delegate_action = SignedDelegateAction {
    delegate_action
    signature,
}
```

There are several ways how the prefix can be added. 
What I think you want, based on this proposal, is this:

```rust
// NEP-366 with your proposal
let delegate_action = DelegateAction {
   prefix: [ 0xFF,0xFF,0xFF,0xFF]; // this would be the prefix
   ... // here go all the other values
};
let signature = sign(delegate_action); // note the this has no hash
let signed_delegate_action = SignedDelegateAction {
    delegate_action
    signature,
}
```

But what feels more natural, at least for meta transactions, is this:
```rust
// NEP-366 with prefix inside hash
let delegate_action = DelegateAction {...};
// prefix is only prefixed for the hash and signature, not in the actual message
let signature = sign(hash( [ 0xFF,0xFF,0xFF,0xFF] +  delegate_action));
let signed_delegate_action = SignedDelegateAction {
    delegate_action
    signature,
}
```
Including the prefix in the message itself seems like unnecessary overhead. A receiver knows from the context what the message is supposed to be. The 4 bytes are unnecessary overhead to send over the wire, store on chain, etc. Therefore I would suggest to write in the NEP that this prefix is added implicitly and doesn't have to be included in the actual struct.

Regarding hashing or not hashing, I am not sure if there is any gain from the extra hash, ed25519 hashes all the bytes anyway. But if we want this proposal to be compatible with meta transactions (or vice-versa) either we should define to use the hash or change the meta transaction implementation which is about to be stabilized any moment now.

Reply:
Indeed, this is a great point, I have changed the prefix to be implicit

Reply:
I meant the hash to still be there, since this is how traditionally signing works in NEAR. I made this more explicit now. Thanks for the feedback!